feat: implement new optimal control example

YousefElbrolosy opened the pull request:
I have created a new folder under tutorials-v5 called quantum optimal control, to include examples of how to apply optimal control algorithms, since many of the linked examples to qutip-notebooks (https://qutip.org/qutip-tutorials/) are now outdated.

[This is an updated implementation using qutip_qtrl based on the deprecated notebook of GRAPE CNOT implementation by Robert Johansson](https://nbviewer.org/github/qutip/qutip-notebooks/blob/master/examples/control-grape-cnot.ipynb)

Was directed here by @hodgestar after a pull request to fix an outdated notebook in https://github.com/qutip/qutip-notebooks

pmenczel commented:
Hi and thank you for contributing. Migrating optimal control notebooks to qutip-tutorials and updating them is very good work!

Please have a look at the [main page of this repository](https://github.com/qutip/qutip-tutorials). This repository does not contain tutorials in ipynb format, but rather in jupytext markdown format. In markdown format, it will be easier to review the contents of the notebook.

@rochisha0 is also migrating optimal control notebooks in PR #122. However, this one here still seems to be missing from the other PR, so that is good. Could you please have a look at the other PR and follow the folder structure there?

YousefElbrolosy commented:
Done!

pmenczel commented:
Thank you! You are adding the same file twice now I think, the "quantum optimal control" one should be deleted.

It seems like the drift Hamiltonian was zero in the old notebook, and is not zero any more here. Could you just briefly explain the changes?

YousefElbrolosy commented:
> Thank you! You are adding the same file twice now I think, the "quantum optimal control" one should be deleted.
> 
> It seems like the drift Hamiltonian was zero in the old notebook, and is not zero any more here. Could you just briefly explain the changes?

You are correct, I was experimenting with it and forgot to change it back.
I fixed it and removed the extra file.

pmenczel commented on tutorials-v5/optimal-control/control-grape-cnot.md:
````diff
@@ -0,0 +1,292 @@
+
+## Fidelity/overlap
+
+```python
+# initially its dimensions where [[2,2],[2,2]]
+U_target = qt.Qobj(U_target.data, dims=[[4], [4]])
+U_target
+```
````
Oh. Well, `[[2,2], [2,2]]` are the correct dimensions here, it seems like a bug in qtrl that the dimensions of the output are wrong. It doesn't really make a difference, but I'd rather set the dimensions of `U_f` to the correct value. That could be done simply with
```
U_f = result.evo_full_final
U_f.dims = [[2,2], [2,2]]
```

pmenczel commented on tutorials-v5/optimal-control/control-grape-cnot.md:
````diff
@@ -0,0 +1,292 @@
+```
+
+```python
+def overlap(U_target, U_f):
+    """
+    Calculate the overlap between the target unitary U_target and
+    the final unitary U_f.
+
+    Parameters:
+    U_target (qutip.Qobj): Target unitary operator.
+    U_f (qutip.Qobj): Final unitary operator.
+
+    Returns:
+    float: Real part of the overlap value.
+    float: Fidelity (absolute square of the overlap).
+    """
+    # dividing over U_target.shape[0] is for normalization
+    overlap_value = (U_target.dag() * U_f).tr() / U_target.shape[0]
+    fidelity = abs(overlap_value) ** 2
+    return overlap_value.real, fidelity
+
+
+# Example usage
+overlap_real, fidelity = overlap(U_target, U_f)
+print(f"Overlap (real part): {overlap_real}")
+print(f"Fidelity: {fidelity}")
````
We can simplify by using a qutip built-in function (which doesn't give the real part of the overlap, but I think that's alright)

```suggestion
print(f"Fidelity: {qt.process_fidelity(U_f, U_target)}")
```

pmenczel commented on tutorials-v5/optimal-control/control-grape-cnot.md:
```diff
@@ -0,0 +1,292 @@
+## Proceess tomography
+
+
+Quantum Process Tomography (QPT) is a technique used to characterize an unknown quantum operation by reconstructing its process matrix (also called the χ (chi) matrix). This matrix describes how an input quantum state is transformed by the operation.
```
Thank you for adding more explanations!

pmenczel replied:
Thank you for adding more explanations!

pmenczel commented on tutorials-v5/optimal-control/control-grape-cnot.md:
````diff
@@ -0,0 +1,292 @@
+
+
+```python
+version_table()
````
```suggestion
qt.about()
```

pmenczel reviewed (commented):
Thank you. I am adding a few suggestions for improvement.

@ajgpitch I see that the other control notebooks that @rochisha0 has converted are numbered now. How did you imagine this, should we just give this one the number 08?

YousefElbrolosy commented:
> Thank you. I am adding a few suggestions for improvement.
> 
> @ajgpitch I see that the other control notebooks that @rochisha0 has converted are numbered now. How did you imagine this, should we just give this one the number 08?

Maybe, when @rochisha0 merge passes the build tests, I can rebase and then modify the rest of the numbering so that all the grape methods are numbered one after the other to maintain natural ordering.

YousefElbrolosy commented:
Any ideas, about fixing this dead links error?

YousefElbrolosy commented:
> > Any ideas, about fixing this dead links error in the CI/CD checks?
> 
> It's a problem that was fixed in another PR... could you merge the current main branch into this one? Then the tests should pass (and if they do, I'll merge :))

I did, and modified the notebook name to be consistent as well could you please try the workflow?

pmenczel commented:
Great, thank you! Reminder @ajgpitch and @rochisha0 , feel free to rename this to fit your naming scheme and add it to the overview when you add the overview back.
